db/db_impl/db_impl_write.cc

```diff
@@ -1437,6 +1439,37 @@ Status DBImpl::ThrottleLowPriWritesIfNeeded(const WriteOptions& write_options,
   return Status::OK();
 }
 
+void DBImpl::MaybeFlushStatsCF(autovector<ColumnFamilyData*>* cfds) {
+  assert(cfds != nullptr);
+  if (!cfds->empty() && immutable_db_options_.persist_stats_to_disk) {
+    ColumnFamilyData* cfd_stats =
+        versions_->GetColumnFamilySet()->GetColumnFamily(
+            kPersistentStatsColumnFamilyName);
+    if (cfd_stats != nullptr && !cfd_stats->mem()->IsEmpty()) {
+      bool stats_cf_flush_needed = true;
```

Comment:
I find `stats_cf_flush_neede` can be a little misleading here. It is possible that the stats cf is full and thus already added to `cfds` by the caller of `MaybeFlushStatsCF`. In this case, even though `stats_cf_flush_needed` will be evaluated as `false`, you still flush the stats cf because it is in `cfds`.
Before iterating through the column family set and comparing log numbers of column families, how about we check whether `stats_cfd` is already in `cfds`. If so, it means we need to flush the stats cf due to memory budget restrictions. Then we should just return. If not, we can do iterate the column family set and see if we should flush the stats cf based on log number comparison.

Reply:
there is already such a check in line 1458, but yes we can optimize the code a bit by returning early

Reply:
There are two things here. First, maybe we can rename `stats_cf_flush_needed` to something else, as explained in my previous comment. Second, we can move the check from line 1458, as you suggest.